altosui/altosui.1
.\"
.\" Copyright © 2010 Bdale Garbee <bdale@gag.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH ALTOSUI 1 "altosui" ""
.SH NAME
altosui \- Rocket flight monitor
.SH SYNOPSIS
.B "altosui"
.SH DESCRIPTION
.I altosui
connects to a TeleDongle or TeleMetrum device through a USB serial device.
It provides a menu-oriented
user interface to monitor, record and review rocket flight data.
.SH USAGE
When connected to a TeleDongle device, altosui turns on the radio
receiver and listens for telemetry packets. It displays the received
telemetry data, and reports flight status via voice synthesis. All
received telemetry information is recorded to a file.
.P
When connected to a TeleMetrum device, altosui can be used to configure the
TeleMetrum, and to downloads the eeprom data and store it in a file.
.P
A number of other menu options exist, including the ability to export flight
data in different formats.
.SH FILES
All data log files are recorded into a user-specified directory
(default ~/TeleMetrum). Files are named using the current date, the serial
number of the reporting device, the flight number recorded in the data
and either '.telem' for telemetry data or '.eeprom' for eeprom data.
.SH AUTHOR
Keith Packard


altos/ChangeLog
commit dbcb68f684a96d13efeb9993685f61b27b45e7dc
Author: Bdale Garbee <bdale@gag.com>
Date:   Sat Aug 12 00:21:43 2017 -0400

    update docinfo file for 1.8 release

commit dfb9568866ba250a02039cc3429d5ca0603a60d9
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Aug 12 00:10:23 2017 -0400

    Note that telebt 4.0 files need to go to altusmetrumllc dir
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 6b083e90d5b8de215f2f303f18b5c4ac9185622d
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Aug 12 00:06:48 2017 -0400

    Bump version to 1.8
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 243c1e519d41b45765d4edf5f608655d6ab01eff
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Aug 12 00:06:26 2017 -0400

    doc: Add 1.8 release notes
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 7b67dcdd90317845261e51a1adb2404dcb0f2507
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Aug 12 00:05:50 2017 -0400

    altosui, telegps: Add TeleBT v4.0 firmware to app packages
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 5fdce5b4a70ae59f8ac1ca10d9fadbbbd62412f3
Merge: f1bbf1bd 2a3db289
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Aug 11 23:45:25 2017 -0400

    Merge remote-tracking branch 'origin/master'

commit f1bbf1bd18d52a084af714a022b12bfc3c76374c
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Aug 11 23:40:41 2017 -0400

    ao-bringup: Allow dfu_util to "fail" while flashing chaoskey
    
    Sometimes it isn't happy about the exit status, but if the device
    comes up and works OK, we don't care.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit fbe87bd18308cab62eda6d5e956a8618095ed33b
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Aug 11 23:39:13 2017 -0400

    altoslib: Set time in state for KML output correctly
    
    Use new computed state_time value.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c7b34e6691b632199bc7802c2e4dea4de992578e
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:49:26 2017 -0700

    libaltos: Test both bluetooth and USB APIs with cjnitest
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 52bbfa9cd46ca27a4de937e89d48246a4e94c8a6
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:47:40 2017 -0700

    altos/test: Fix AES key format in test code
    
    Use temp variable instead of stepping on the AES name.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 9eaf917707287af57c2dadfc82d6d415b0964329
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:46:24 2017 -0700

    altos/test: Fix printf format for int64 arithmetic test
    
    Use 'long long' and %lld for 64-bit values when printing.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 198a3d4d5f51de5c0f0413299582479cde4e177f
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:37:27 2017 -0700

    altosui: Rename AltosConfig -> AltosConfigFC
    
    Now the class name says that this configures the flight computer.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c7c2fc2d85414fefda0a7948a6c4e38f65140861
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:38:09 2017 -0700

    altosui: Adapt to flight stats time value changes
    
    There aren't state time values anymore as those don't work when you
    have multiple motors. Instead, 'boost_time' is when the rocket left
    the pad and 'landed_time' is when it touched down. Use these new
    values in the --summary output.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 2a3db28953f5cd981c519a917cdcd9f48dcb08e9
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:49:21 2017 -0600

    add TeleBT v4.0 to the list of stuff we build

commit 04560a7cd6a64e742954a2d0fbd7563e2c310402
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:46:55 2017 -0600

    ignore more build products in telebt-v4.0 tree

commit c8dbcaf69cd538a31ab6e2b568237ae7c8656a9a
Merge: 0cbfa444 ea3b5815
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:44:03 2017 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 0cbfa444a9f9159cb509bb47ca5590fc1d709f64
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:43:38 2017 -0600

    clean up telefireone name references

commit 3a1512b605a81ebee23cc6b7980b709a54cb43c9
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:42:54 2017 -0600

    rename existing turnon_telebt script, create a new one for v4.0

commit ea3b5815b27005b2f4c3034715f656d28ea8534e
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Jul 21 16:16:55 2017 -0700

    libaltos: Add btle test program
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 545bdc1f8046b08e237307d97d0044ddae9095b8
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Jul 21 16:14:31 2017 -0700

    altos/test: Add EasyMini test mode
    
    Use baro-only mode, parse easymini logs.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 2969f0af52dfbdea7593b208e510c7b17227064c
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Jul 21 16:09:55 2017 -0700

    altoslib: fix adjustment of landed time
    
    Push landed time out if we computed silliness.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 30b308ad0189b4aa9dd5cc7ee876b55fee355a56
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Jul 21 16:05:45 2017 -0700

    Set version to 1.7.1 for TeleBT firmware build
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit b923e25a8a49dda40e62070dd2df6ff9130d4b38

altos/altosui/AltosUI.java
/*
 * Copyright © 2010 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

package altosui;

import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
import java.io.*;
import java.util.concurrent.*;
import org.altusmetrum.altoslib_14.*;
import org.altusmetrum.altosuilib_14.*;

public class AltosUI extends AltosUIFrame implements AltosEepromGrapher {
	public AltosVoice voice = new AltosVoice();

	public static boolean load_library(Frame frame) {
		if (!Altos.load_library()) {
			JOptionPane.showMessageDialog(frame,
						      String.format("No AltOS library in \"%s\"",
								    System.getProperty("java.library.path","<undefined>")),
						      "Cannot load device access library",
						      JOptionPane.ERROR_MESSAGE);
			return false;
		}
		return true;
	}

	void telemetry_window(AltosDevice device) {
		try {
			AltosFlightReader reader = new AltosTelemetryReader(new AltosSerial(device));
			if (reader != null)
				new AltosFlightUI(voice, reader, device.getSerial());
		} catch (FileNotFoundException ee) {
			JOptionPane.showMessageDialog(AltosUI.this,
						      ee.getMessage(),
						      String.format ("Cannot open %s", device.toShortString()),
						      JOptionPane.ERROR_MESSAGE);
		} catch (AltosSerialInUseException si) {
			JOptionPane.showMessageDialog(AltosUI.this,
						      String.format("Device \"%s\" already in use",
								    device.toShortString()),
						      "Device in use",
						      JOptionPane.ERROR_MESSAGE);
		} catch (IOException ee) {
			JOptionPane.showMessageDialog(AltosUI.this,
						      String.format ("Unknown I/O error on %s", device.toShortString()),
						      "Unknown I/O error",
						      JOptionPane.ERROR_MESSAGE);
		} catch (TimeoutException te) {
			JOptionPane.showMessageDialog(this,
						      String.format ("Timeout on %s", device.toShortString()),
						      "Timeout error",
						      JOptionPane.ERROR_MESSAGE);
		} catch (InterruptedException ie) {
			JOptionPane.showMessageDialog(this,
						      String.format("Interrupted %s", device.toShortString()),
						      "Interrupted exception",
						      JOptionPane.ERROR_MESSAGE);
		}
	}

	public void scan_device_selected(AltosDevice device) {
		telemetry_window(device);
	}

	Container	pane;
	GridBagLayout	gridbag;

	JButton addButton(int x, int y, String label) {
		GridBagConstraints	c;
		JButton			b;

		c = new GridBagConstraints();
		c.gridx = x; c.gridy = y;
		c.fill = GridBagConstraints.BOTH;
		c.weightx = 1;
		c.weighty = 1;
		b = new JButton(label);

		//Dimension ps = b.getPreferredSize();

		gridbag.setConstraints(b, c);
		add(b, c);
		return b;
	}

	/* OSXAdapter interfaces */
	public void macosx_file_handler(String path) {
		process_graph(null, new File(path));
	}

	public void macosx_quit_handler() {
		System.exit(0);
	}

	public void macosx_preferences_handler() {
		ConfigureAltosUI();
	}

	public AltosUI() {

		load_library(null);

		register_for_macosx_events();

		AltosUIPreferences.set_component(this);

		pane = getContentPane();
		gridbag = new GridBagLayout();
		pane.setLayout(gridbag);

		JButton	b;

		b = addButton(0, 0, "Monitor Flight");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						ConnectToDevice();
					}
				});
		b.setToolTipText("Connect to TeleDongle and monitor telemetry");
		b = addButton(1, 0, "Save Flight Data");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						SaveFlightData();
					}
				});
		b.setToolTipText("Download and/or delete flight data from an altimeter");
		b = addButton(2, 0, "Replay Flight");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						Replay();
					}
				});
		b.setToolTipText("Watch an old flight in real-time");
		b = addButton(3, 0, "Graph Data");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						GraphData();
					}
				});
		b.setToolTipText("Present flight data in a graph and table of statistics");
		b = addButton(4, 0, "Export Data");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						ExportData();
					}
				});
		b.setToolTipText("Convert flight data for a spreadsheet or GoogleEarth");
		b = addButton(0, 1, "Configure Altimeter");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						ConfigureTeleMetrum();
					}
				});
		b.setToolTipText("Set flight, storage and communication parameters");
		b = addButton(1, 1, "Configure AltosUI");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					ConfigureAltosUI();
				}
			});
		b.setToolTipText("Global AltosUI settings");

		b = addButton(2, 1, "Configure Ground Station");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					ConfigureTeleDongle();
				}
			});

		b = addButton(3, 1, "Flash Image");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					FlashImage();
				}
			});
		b.setToolTipText("Replace the firmware in any AltusMetrum product");

		b = addButton(4, 1, "Fire Igniter");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					FireIgniter();
				}
			});
		b.setToolTipText("Remote control of igniters for deployment testing");
		b = addButton(0, 2, "Scan Channels");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					ScanChannels();
				}
			});
		b.setToolTipText("Find what channel an altimeter is sending telemetry on");
		b = addButton(1, 2, "Load Maps");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					LoadMaps();
				}
			});
		b.setToolTipText("Download satellite images for off-line flight monitoring");
		b = addButton(2, 2, "Monitor Idle");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					IdleMonitor();
				}
			});
		b.setToolTipText("Check flight readiness of altimeter in idle mode");

//		b = addButton(3, 2, "Launch Controller");
//		b.addActionListener(new ActionListener() {
//				public void actionPerformed(ActionEvent e) {
//					LaunchController();
//				}
//			});

		b = addButton(4, 2, "Quit");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					System.exit(0);
				}
			});
		b.setToolTipText("Close all active windows and terminate AltosUI");

		setTitle("AltOS");

		pane.doLayout();
		pane.validate();

		doLayout();
		validate();

		setDefaultCloseOperation(JFrame.DO_NOTHING_ON_CLOSE);
		addWindowListener(new WindowAdapter() {
			@Override
			public void windowClosing(WindowEvent e) {
				System.exit(0);
			}
		});

		setLocationByPlatform(false);

		/* Insets aren't set before the window is visible */
		setVisible(true);
	}

	private void ConnectToDevice() {
		AltosDevice	device = AltosDeviceUIDialog.show(AltosUI.this,
								Altos.product_basestation);

		if (device != null)
			telemetry_window(device);
	}

	void ConfigureCallsign() {
		String	result;
		result = JOptionPane.showInputDialog(AltosUI.this,
						     "Configure Callsign",
						     AltosUIPreferences.callsign());
		if (result != null)
			AltosUIPreferences.set_callsign(result);
	}

	void ConfigureTeleMetrum() {
		new AltosConfigFC(AltosUI.this);
	}

	void ConfigureTeleDongle() {
		new AltosConfigTD(AltosUI.this);
	}

	void FlashImage() {
		AltosFlashUI.show(AltosUI.this);
	}

	void FireIgniter() {
		new AltosIgniteUI(AltosUI.this);
	}

	void ScanChannels() {
		new AltosScanUI(AltosUI.this, true);
	}

	void LoadMaps() {
		new AltosUIMapPreload(AltosUI.this);
	}

	void LaunchController() {
		new AltosLaunchUI(AltosUI.this);
	}

	/*
	 * Replay a flight from telemetry data
	 */
	private void Replay() {
		AltosDataChooser chooser = new AltosDataChooser(
			AltosUI.this);

		AltosRecordSet set = chooser.runDialog();
		if (set != null) {
			AltosReplayReader reader = new AltosReplayReader(set, chooser.file());
			new AltosFlightUI(voice, reader);
		}
	}

	/* Connect to TeleMetrum, either directly or through
	 * a TeleDongle over the packet link
	 */

	public void graph_flights(AltosEepromList flights) {
		for (AltosEepromLog flight : flights) {
			if (flight.graph_selected && flight.file != null) {
				process_graph(this, flight.file);
			}
		}
	}

	private void SaveFlightData() {
		new AltosEepromManage(this, this, AltosLib.product_any);
	}

	private static AltosFlightSeries make_series(AltosRecordSet set) {
		AltosFlightSeries series = new AltosFlightSeries(set.cal_data());
		set.capture_series(series);
		series.finish();
		return series;
	}

	/* Load a flight log file and write out a CSV file containing
	 * all of the data in standard units
	 */

	private void ExportData() {
		AltosDataChooser chooser;
		chooser = new AltosDataChooser(this);
		AltosRecordSet set = chooser.runDialog();
		if (set == null)
			return;
		AltosFlightSeries series = make_series(set);
		new AltosCSVUI(AltosUI.this, series, chooser.file());
	}

	private static boolean graph_file(AltosUI altosui, AltosRecordSet set, File file) {
		if (set == null)
			return false;
		if (!set.valid()) {
			JOptionPane.showMessageDialog(altosui,
						      String.format("Failed to parse file %s", file),
						      "Graph Failed",
						      JOptionPane.ERROR_MESSAGE);
			return false;
		}
		try {
			new AltosGraphUI(set, file);
			return true;
		} catch (InterruptedException ie) {
		} catch (IOException ie) {
		}
		return false;
	}

	/* Load a flight log CSV file and display a pretty graph.
	 */

	private void GraphData() {
		AltosDataChooser chooser;
		chooser = new AltosDataChooser(this);
		AltosRecordSet set = chooser.runDialog();
		graph_file(this, set, chooser.file());
	}

	private void ConfigureAltosUI() {
		new AltosConfigureUI(AltosUI.this, voice);
	}

	private void IdleMonitor() {
		try {
			new AltosIdleMonitorUI(this);
		} catch (Exception e) {
		}
	}

	static AltosWriter open_csv(File file) {
		try {
			return new AltosCSV(file);
		} catch (FileNotFoundException fe) {
			System.out.printf("%s\n", fe.getMessage());
			return null;
		}
	}

	static AltosWriter open_kml(File file) {
		try {
			return new AltosKML(file);
		} catch (FileNotFoundException fe) {
			System.out.printf("%s\n", fe.getMessage());
			return null;
		}
	}

	static AltosRecordSet record_set(File input) {
		try {
			return AltosLib.record_set(input);
		} catch (IOException ie) {
			String message = ie.getMessage();
			if (message == null)
				message = String.format("%s (I/O error)", input.toString());
			System.err.printf("%s: %s\n", input.toString(), message);
		}
		return null;
	}

	static final int process_none = 0;
	static final int process_csv = 1;
	static final int process_kml = 2;
	static final int process_graph = 3;
	static final int process_replay = 4;
	static final int process_summary = 5;
	static final int process_oneline = 6;

	static boolean process_csv(File input) {
		AltosRecordSet set = record_set(input);
		if (set == null)
			return false;

		File output = Altos.replace_extension(input,".csv");
		System.out.printf("Processing \"%s\" to \"%s\"\n", input, output);
		if (input.equals(output)) {
			System.out.printf("Not processing '%s'\n", input);
			return false;
		} else {
			AltosWriter writer = open_csv(output);
			if (writer == null)
				return false;
			AltosFlightSeries series = make_series(set);
			writer.write(series);
			writer.close();
		}
		return true;
	}

	static boolean process_kml(File input) {
		AltosRecordSet set = record_set(input);
		if (set == null)
			return false;

		File output = Altos.replace_extension(input,".kml");
		System.out.printf("Processing \"%s\" to \"%s\"\n", input, output);
		if (input.equals(output)) {
			System.out.printf("Not processing '%s'\n", input);
			return false;
		} else {
			AltosWriter writer = open_kml(output);
			if (writer == null)
				return false;
			AltosFlightSeries series = make_series(set);
			series.finish();
			writer.write(series);
			writer.close();
			return true;
		}
	}

	static AltosReplayReader replay_file(File file) {
		AltosRecordSet set = record_set(file);
		if (set == null)
			return null;
		return new AltosReplayReader(set, file);
	}

	static boolean process_replay(File file) {
		AltosReplayReader reader = replay_file(file);
		if (reader == null)
			return false;
		AltosFlightUI flight_ui = new AltosFlightUI(new AltosVoice(), reader);
		return true;
	}

	static boolean process_graph(AltosUI altosui, File file) {
		AltosRecordSet set = record_set(file);
		return graph_file(altosui, set, file);
	}

	static boolean process_summary(File file) {
		AltosRecordSet set = record_set(file);
		if (set == null)
			return false;
		System.out.printf("%s:\n", file.toString());
		AltosFlightSeries series = make_series(set);
		AltosFlightStats stats = new AltosFlightStats(series);
		if (stats.serial != AltosLib.MISSING)
			System.out.printf("Serial:       %5d\n", stats.serial);
		if (stats.flight != AltosLib.MISSING)
			System.out.printf("Flight:       %5d\n", stats.flight);
		if (stats.year != AltosLib.MISSING)
			System.out.printf("Date:    %04d-%02d-%02d\n",
					  stats.year, stats.month, stats.day);
		if (stats.hour != AltosLib.MISSING)
			System.out.printf("Time:      %02d:%02d:%02d UTC\n",
					  stats.hour, stats.minute, stats.second);
		if (stats.max_height != AltosLib.MISSING)
			System.out.printf("Max height:  %6.0f m    %6.0f ft\n",
					  stats.max_height,
					  AltosConvert.meters_to_feet(stats.max_height));
		if (stats.max_speed != AltosLib.MISSING)
			System.out.printf("Max speed:   %6.0f m/s  %6.0f ft/s  %6.4f Mach\n",
					  stats.max_speed,
					  AltosConvert.meters_to_feet(stats.max_speed),
					  AltosConvert.meters_to_mach(stats.max_speed));
		if (stats.max_acceleration != AltosLib.MISSING) {
			System.out.printf("Max accel:   %6.0f m/s² %6.0f ft/s² %6.2f g\n",
					  stats.max_acceleration,
					  AltosConvert.meters_to_feet(stats.max_acceleration),
					  AltosConvert.meters_to_g(stats.max_acceleration));
		}
		if (stats.state_speed[Altos.ao_flight_drogue] != AltosLib.MISSING)
			System.out.printf("Drogue rate: %6.0f m/s  %6.0f ft/s\n",
					  stats.state_speed[Altos.ao_flight_drogue],
					  AltosConvert.meters_to_feet(stats.state_speed[Altos.ao_flight_drogue]));
		if (stats.state_speed[Altos.ao_flight_main] != AltosLib.MISSING)
			System.out.printf("Main rate:   %6.0f m/s  %6.0f ft/s\n",
					  stats.state_speed[Altos.ao_flight_main],
					  AltosConvert.meters_to_feet(stats.state_speed[Altos.ao_flight_main]));
		if (stats.landed_time != AltosLib.MISSING &&
		    stats.boost_time != AltosLib.MISSING &&
		    stats.landed_time > stats.boost_time)
			System.out.printf("Flight time: %6.0f s\n",
					  stats.landed_time -
					  stats.boost_time);
		System.out.printf("\n");
		return true;
	}

	static boolean process_oneline(File file) {
		AltosRecordSet set = record_set(file);
		if (set == null)
			return false;
		System.out.printf("%s", file.toString());
		AltosFlightSeries series = make_series(set);
		AltosFlightStats stats = new AltosFlightStats(series);
		if (stats.max_height != AltosLib.MISSING)
			System.out.printf(" height  %6.0f m", stats.max_height);
		if (stats.max_speed != AltosLib.MISSING)
			System.out.printf(" speed   %6.0f m/s", stats.max_speed);
		if (stats.state_enter_speed[AltosLib.ao_flight_drogue] != AltosLib.MISSING)
			System.out.printf(" drogue-deploy   %6.0f m/s", stats.state_enter_speed[AltosLib.ao_flight_drogue]);
		if (stats.max_acceleration != AltosLib.MISSING)
			System.out.printf(" accel   %6.0f m/s²", stats.max_acceleration);
		System.out.printf("\n");
		return true;
	}

	public static void help(int code) {
		System.out.printf("Usage: altosui [OPTION]... [FILE]...\n");
		System.out.printf("  Options:\n");
		System.out.printf("    --replay <filename>\t\trelive the glory of past flights \n");
		System.out.printf("    --graph <filename>\t\tgraph a flight\n");
		System.out.printf("    --summary <filename>\t\tText summary of a flight\n");
		System.out.printf("    --oneline <filename>\t\tOne line summary of a flight\n");
		System.out.printf("    --csv\tgenerate comma separated output for spreadsheets, etc\n");
		System.out.printf("    --kml\tgenerate KML output for use with Google Earth\n");
		System.exit(code);
	}

	public static void main(final String[] args) {
		int	errors = 0;
		load_library(null);
		try {
			UIManager.setLookAndFeel(AltosUIPreferences.look_and_feel());
		} catch (Exception e) {
		}
		AltosUI altosui = null;

		/* Handle batch-mode */
		if (args.length == 0) {
			altosui = new AltosUI();
			java.util.List<AltosDevice> devices = AltosUSBDevice.list(Altos.product_basestation);
			if (devices != null)
				for (AltosDevice device : devices)
					altosui.telemetry_window(device);
		} else {
			int process = process_none;
			for (int i = 0; i < args.length; i++) {
				if (args[i].equals("--help"))
					help(0);
				else if (args[i].equals("--replay"))
					process = process_replay;
				else if (args[i].equals("--kml"))
					process = process_kml;
				else if (args[i].equals("--csv"))
					process = process_csv;
				else if (args[i].equals("--graph"))
					process = process_graph;
				else if (args[i].equals("--summary"))
					process = process_summary;
				else if (args[i].equals("--oneline"))
					process = process_oneline;
				else if (args[i].startsWith("--"))
					help(1);
				else {
					File file = new File(args[i]);
					switch (process) {
					case process_none:
						if (altosui == null)
							altosui = new AltosUI();
					case process_graph:
						if (!process_graph(null, file))
							++errors;
						break;
					case process_replay:
						if (!process_replay(file))
							++errors;
						break;
					case process_kml:
						if (!process_kml(file))
							++errors;
						break;
					case process_csv:
						if (!process_csv(file))
							++errors;
						break;
					case process_summary:
						if (!process_summary(file))
							++errors;
						break;
					case process_oneline:
						if (!process_oneline(file))
							++errors;
						break;
					}
				}
			}
		}
		if (errors != 0)
			System.exit(errors);
	}
}
